source/Calamari.Tests/AWS/S3Fixture.cs

```diff
@@ -293,6 +293,198 @@ public async Task UploadPackageWithMetadata()
+        {
+            var fileName = $"{packageId}.{packageVersion}.{packageExtension}";
+            
+            var packageOptions = new List<S3PackageOptions>
+            {
+                new S3PackageOptions
+                {
+                    StorageClass = "STANDARD",
+                    CannedAcl = "private",
+                    StructuredVariableSubstitutionPatterns = "*.json",
+                    BucketKeyBehaviour = BucketKeyBehaviourType.Filename,
+                }
+            };
+
+            var variables = new CalamariVariables();
+            variables.Set("Property1:Property2:Value", "InjectedValue");
+            
+            var packageFilePath = TestEnvironment.GetTestPath("AWS", "S3", "CompressedPackages", fileName);
+
+            var prefix = UploadEntireCompressedPackage(packageFilePath, packageId, packageVersion, packageOptions, variables);
+
+            await Validate(async client =>
+                           {
+                               var file = await client.GetObjectAsync(bucketName, $"{prefix}{fileName}");
+                               var memoryStream = new MemoryStream();
+                               await file.ResponseStream.CopyToAsync(memoryStream);
+                               var text = await new StreamReader(ZipArchive.Open(memoryStream).Entries.First(entry => entry.Key == "file.json").OpenEntryStream()).ReadToEndAsync();
+                               JObject.Parse(text)["Property1"]["Property2"]["Value"].ToString().Should().Be("InjectedValue");
+                               memoryStream.Close();
+                           });
+        }
+        
+        [Test]
+        public async Task SubstituteVariablesAndUploadTarArchives()
+        {
+            const string packageId = "TestTarPackage";
+            const string packageVersion = "0.0.1";
+            const string packageExtension = "tar";
+            var fileName = $"{packageId}.{packageVersion}.{packageExtension}";
+            
+            var packageOptions = new List<S3PackageOptions>
+            {
+                new S3PackageOptions
+                {
+                    StorageClass = "STANDARD",
+                    CannedAcl = "private",
+                    StructuredVariableSubstitutionPatterns = "*.json",
+                    BucketKeyBehaviour = BucketKeyBehaviourType.Filename,
+                }
+            };
+
+            var variables = new CalamariVariables();
+            variables.Set("Property1:Property2:Value", "InjectedValue");
+            
+            var packageFilePath = TestEnvironment.GetTestPath("AWS", "S3", "CompressedPackages", fileName);
+
+            var prefix = UploadEntireCompressedPackage(packageFilePath, packageId, packageVersion, packageOptions, variables);
+
+            await Validate(async client =>
+                           {
+                               var file = await client.GetObjectAsync(bucketName, $"{prefix}{fileName}");
```

Comment:
This pattern of downloading to a temp file is a bit repetitive and verbose (it's distracting from what's actually being tested). It might be worth extracting the pattern into a helper?